### ISSUE TITLE
Missing destructor call after assignment

### ISSUE THREAD

**Issue by rbberger** [llvm/llvm-project#62818]
Here is a minimal example that reproduces the issue: https://godbolt.org/z/EqdrWznfe

```c++
// DO_SOMETHING from the A destructor should be called when calling test()

void DO_SOMETHING();

struct A {
  A() {};
  // workaround 1
  // issue goes away with
  // A() = default;
  ~A() {
    DO_SOMETHING();
  }

  A & operator=(A a) & {
    return *this;
  }
};

// workaround 2: issue goes away if B is not a template
template<typename T>
struct B {
  void test() {
    // workaround 3
    //#ifdef __clang__
    //this->
    //#endif
    a = {};

    // workaround 4
    //a = decltype(a)();
  }

private:
  A a;
};

// look at disassembly of this function, it should include a call to DO_SOMETHING
void client(B<int> &f) {f.test();}
```

In essence, it assigns `{}` to a member variable, which should trigger the destruction of the temporary materialized to call the assignment operator.
The disassembly should include a call to  `DO_SOMETHING`, which it does for GCC, but not Clang. The example shows 4 possible workarounds, including changing `a = {}` to `this->a = {}` which then produces the expected call in disassembly.

**Comment by shafik** [comment]
This looks like a regression from clang-15: https://godbolt.org/z/PjK348Ph1

CC'ing folks who might have an idea of the change that caused this: @alanzhao1 @zygoloid @royjacobson @erichkeane @cor3ntin 

I don't believe this is an opt issue, if we look at the AST we can see we are not generating an `ExprWithCleanups`

**Comment by opensdh** [comment]
> This looks like a regression from clang-15: https://godbolt.org/z/PjK348Ph1

Our original code failed identically in 13.0.1, in case that broadens the search usefully.

**Comment by shafik** [comment]
> > This looks like a regression from clang-15: https://godbolt.org/z/PjK348Ph1
> 
> Our original code failed identically in 13.0.1, in case that broadens the search usefully.

Do you have an example that fails in 13.0.1? This case does not seem to demonstrate the issue for this code: https://godbolt.org/z/E5dd161dP

Was there a bug report against 13.0.1?

**Comment by rbberger** [comment]
@shafik It does happen with 13.0.1, but you need to add `-std=c++17`.

**Comment by shafik** [comment]
> @shafik It does happen with 13.0.1, but you need to add `-std=c++17`.

That is super helpful, this is change in behavior from C++14 to C++17: https://godbolt.org/z/T3a65b8aW

It exists further back if we use C++17.

**Comment by royjacobson** [comment]
I have no idea why it depends on being a template, but since C++17 did add copy elision for assignments the behavior might be legal.

**Comment by Endilll** [comment]
I don't think it's legal, because copy elision is limited to return statements, throw expressions, expression declarations, and (since [C++20](http://eel.is/c++draft/class.copy.elision#1.3)) coroutines parameters ([C++17](https://timsong-cpp.github.io/cppwp/n4659/class.copy.elision#1)). I don't see how `B::test()` and `client()` fall under any of those conditions.

**Comment by royjacobson** [comment]
Yeah, you're right.

From looking at the AST it seems the `ExprWithCleanups` node is dropped for some reason in the instantiation of the function.

**Comment by zygoloid** [comment]
Most likely, we're transforming the RHS of an assignment expression using `TransformExpr` instead of `TransformInitializer`, so we're not properly reverting the initializer list to its syntactic form before handing it to `Sema`.

**Comment by shafik** [comment]
After digging into this a bit I think this may be similar to https://reviews.llvm.org/D148474 in that we are not filtering when we apply guaranteed copy elision well enough. I am not sure if this should apply in the case of `InitializedEntity::EK_Parameter`.

**Comment by shafik** [comment]
> After digging into this a bit I think this may be similar to https://reviews.llvm.org/D148474 in that we are not filtering when we apply guaranteed copy elision well enough. I am not sure if this should apply in the case of `InitializedEntity::EK_Parameter`.

Ok, filtering on that in `TryConstructorInitialization(...)` fixes this case but breaks a lot of other stuff, oh well. Either I need a better filter or that is not the right place to fix this.

**Comment by zygoloid** [comment]
Looks like the bad calls to `TransformExpr` are [here](https://github.com/llvm/llvm-project/blob/3b272039a88cc9952e5bad71b6b02a62b2c22098/clang/lib/Sema/TreeTransform.h#L11405) and [here](https://github.com/llvm/llvm-project/blob/3b272039a88cc9952e5bad71b6b02a62b2c22098/clang/lib/Sema/TreeTransform.h#L11953). For an assignment operator, we can have an initializer list as the syntactic second operand, so we need to use `TransformInitializer` instead in at least those cases (and it's probably correct to do so unconditionally).

**Comment by shafik** [comment]
> Looks like the bad calls to `TransformExpr` are [here](https://github.com/llvm/llvm-project/blob/3b272039a88cc9952e5bad71b6b02a62b2c22098/clang/lib/Sema/TreeTransform.h#L11405) and [here](https://github.com/llvm/llvm-project/blob/3b272039a88cc9952e5bad71b6b02a62b2c22098/clang/lib/Sema/TreeTransform.h#L11953). For an assignment operator, we can have an initializer list as the syntactic second operand, so we need to use `TransformInitializer` instead in at least those cases (and it's probably correct to do so unconditionally).

This looks like it fixes this issue, I will try and post a PR tomorrow.

**Comment by shafik** [comment]
PR: https://reviews.llvm.org/D151235

CC @zygoloid
